File: jde-mcp-ube-analyzer/src/tools/integrity.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import {
  JdeApVoucherQuerySchema,
  JdeGlBalanceQuerySchema,
  JdeGlDetailQuerySchema,
  JdeApGlIntegrityCheckSchema,
  type JdeApVoucherQueryInput,
  type JdeGlBalanceQueryInput,
  type JdeGlDetailQueryInput,
  type JdeApGlIntegrityCheckInput,
} from "../schemas/tools.js";
import { queryTable } from "../services/ais-client.js";
import { CHARACTER_LIMIT } from "../constants.js";
import type { AisOperator } from "../types.js";

// ──────────────────────────────────────────────────────────────
// Helpers (same pattern as domain.ts)
// ──────────────────────────────────────────────────────────────

interface FilterDef {
  column: string;
  operator: AisOperator;
  value: string | string[];
}

function addFilter(
  filters: FilterDef[],
  column: string,
  operator: AisOperator,
  value: string | number | undefined
): void {
  if (value === undefined || value === "") return;
  filters.push({ column, operator, value: String(value) });
}

function truncate(text: string, records?: number): string {
  if (text.length <= CHARACTER_LIMIT) return text;
  return (
    text.slice(0, CHARACTER_LIMIT) +
    `\n\n... [TRUNCATED — ${records ?? "many"} total records. Narrow your filters.]`
  );
}

// ──────────────────────────────────────────────────────────────
// Tool 1: A/P Voucher Query (F0411)
// ──────────────────────────────────────────────────────────────

export function registerApVoucherQuery(server: McpServer): void {
  server.registerTool(
    "jde_ap_voucher_query",
    {
      title: "Query A/P Vouchers",
      description: `Query the JDE Accounts Payable Ledger (F0411) for voucher pay items.

Use this to look up AP vouchers filtered by company, supplier, GL offset code (GLPT),
payment status, date range, fiscal year, and period. Essential for R047001A integrity analysis.

Key columns returned: DOC, DCT, KCO, SFX, CO, AN8, MCU, AG (gross), AAP (open), PAAP (original),
GLPT (GL offset), AID, FY, PN, DGJ, PST, VINV.

The GLPT (GL Offset) field is critical — it determines which GL offset account the voucher posts to,
and is the primary grouping key for the R047001A integrity report.

Args:
  - company: Company code (CO)
  - supplierNumber: Supplier address book number (AN8)
  - documentType: Document type (DCT), e.g. 'PV'
  - payStatus: Payment status (PST): A=Approved, P=Paid, V=Void
  - glOffset: GL posting code (GLPT) — the key R047001A grouping field
  - dateFrom/dateTo: GL date range (YYYY-MM-DD)
  - fiscalYear: Fiscal year (FY)
  - period: GL period (PN)
  - maxRows: Max rows (default 100)`,
      inputSchema: JdeApVoucherQuerySchema,
      annotations: {
        readOnlyHint: true,
        destructiveHint: false,
        idempotentHint: true,
        openWorldHint: true,
      },
    },
    async (params: JdeApVoucherQueryInput) => {
      try {
        const filters: FilterDef[] = [];
        addFilter(filters, "CO", "EQUAL", params.company);
        addFilter(filters, "AN8", "EQUAL", params.supplierNumber);
        addFilter(filters, "DCT", "EQUAL", params.documentType);
        addFilter(filters, "PST", "EQUAL", params.payStatus);
        addFilter(filters, "GLPT", "EQUAL", params.glOffset);
        addFilter(filters, "FY", "EQUAL", params.fiscalYear);
        addFilter(filters, "PN", "EQUAL", params.period);
        if (params.dateFrom) addFilter(filters, "DGJ", "GREATER_EQUAL", params.dateFrom);
        if (params.dateTo) addFilter(filters, "DGJ", "LESS_EQUAL", params.dateTo);

        const response = await queryTable({
          tableName: "F0411",
          columns: ["DOC", "DCT", "KCO", "SFX", "CO", "AN8", "MCU", "AG", "AAP", "PAAP", "GLPT", "AID", "FY", "PN", "DGJ", "PST", "VINV"],
          filters: filters.length > 0 ? filters : undefined,
          maxRows: params.maxRows,
        });

        const gridData = response.fs_DATABROWSE?.data?.gridData;
        if (!gridData) {
          return {
            content: [{ type: "text" as const, text: "No vouchers found matching the criteria." }],
          };
        }

        const output = {
          tableName: "F0411",
          description: "A/P Voucher Pay Items",
          totalRecords: gridData.summary.records,
          returnedRecords: gridData.rowset.length,
          hasMore: gridData.summary.moreRecords,
          rows: gridData.rowset,
        };

        return {
          content: [{ type: "text" as const, text: truncate(JSON.stringify(output, null, 2), gridData.summary.records) }],
        };
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        return {
          isError: true,
          content: [{ type: "text" as const, text: `Error querying F0411: ${msg}` }],
        };
      }
    }
  );
}

// ──────────────────────────────────────────────────────────────
// Tool 2: GL Balance Query (F0902)
// ──────────────────────────────────────────────────────────────

export function registerGlBalanceQuery(server: McpServer): void {
  server.registerTool(
    "jde_gl_balance_query",
    {
      title: "Query GL Account Balances",
      description: `Query the JDE Account Balances table (F0902) for period-level GL balances.

Use this to get the GL side of an integrity check. Returns balances by period for each
account/ledger type/fiscal year combination.

Key columns: AID, CO, MCU, OBJ, SUB, LT, FY, CTRY, AN01-AN14 (period amounts), BORG (beginning balance).

For R047001A: query by the object account that corresponds to the GLPT offset, then compare
the period totals against the F0411 voucher sums for the same period/company.

Args:
  - company: Company code (CO)
  - businessUnit: Business unit (MCU)
  - objectAccount: Object account (OBJ), e.g. '1110' for AP Trade
  - subsidiary: Subsidiary (SUB) for detail-level
  - ledgerType: Ledger type (default 'AA' = Actual)
  - fiscalYear: Fiscal year (FY)
  - maxRows: Max rows (default 100)`,
      inputSchema: JdeGlBalanceQuerySchema,
      annotations: {
        readOnlyHint: true,
        destructiveHint: false,
        idempotentHint: true,
        openWorldHint: true,
      },
    },
    async (params: JdeGlBalanceQueryInput) => {
      try {
        const filters: FilterDef[] = [];
        addFilter(filters, "CO", "EQUAL", params.company);
        addFilter(filters, "MCU", "EQUAL", params.businessUnit);
        addFilter(filters, "OBJ", "EQUAL", params.objectAccount);
        addFilter(filters, "SUB", "EQUAL", params.subsidiary);
        addFilter(filters, "LT", "EQUAL", params.ledgerType);
        addFilter(filters, "FY", "EQUAL", params.fiscalYear);

        const response = await queryTable({
          tableName: "F0902",
          columns: ["AID", "CO", "MCU", "OBJ", "SUB", "LT", "FY", "CTRY",
                     "AN01", "AN02", "AN03", "AN04", "AN05", "AN06",
                     "AN07", "AN08", "AN09", "AN10", "AN11", "AN12",
                     "AN13", "AN14", "BORG"],
          filters: filters.length > 0 ? filters : undefined,
          maxRows: params.maxRows,
        });

        const gridData = response.fs_DATABROWSE?.data?.gridData;
        if (!gridData) {
          return {
            content: [{ type: "text" as const, text: "No account balances found matching the criteria." }],
          };
        }

        const output = {
          tableName: "F0902",
          description: "Account Balances by Period",
          totalRecords: gridData.summary.records,
          returnedRecords: gridData.rowset.length,
          hasMore: gridData.summary.moreRecords,
          rows: gridData.rowset,
        };

        return {
          content: [{ type: "text" as const, text: truncate(JSON.stringify(output, null, 2), gridData.summary.records) }],
        };
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        return {
          isError: true,
          content: [{ type: "text" as const, text: `Error querying F0902: ${msg}` }],
        };
      }
    }
  );
}

// ──────────────────────────────────────────────────────────────
// Tool 3: GL Detail / Journal Entry Query (F0901)
// ──────────────────────────────────────────────────────────────

export function registerGlDetailQuery(server: McpServer): void {
  server.registerTool(
    "jde_gl_detail_query",
    {
      title: "Query GL Journal Entries",
      description: `Query the JDE Account Ledger (F0901) for individual GL journal entries.

Use this for drill-down when an integrity discrepancy is found. Shows the individual
transactions that make up the GL balance for a specific account.

Key columns: AID, DOC, DCT, KCO, CO, MCU, LT, FY, PN, AA (amount), DGJ, EXR, AN8, JELN.

Args:
  - accountId: Account ID (AID), e.g. '00001.1110.ACME'
  - company: Company code (CO)
  - documentType: Document type (DCT)
  - fiscalYear: Fiscal year (FY)
  - period: GL period (PN)
  - dateFrom/dateTo: GL date range
  - maxRows: Max rows (default 100)`,
      inputSchema: JdeGlDetailQuerySchema,
      annotations: {
        readOnlyHint: true,
        destructiveHint: false,
        idempotentHint: true,
        openWorldHint: true,
      },
    },
    async (params: JdeGlDetailQueryInput) => {
      try {
        const filters: FilterDef[] = [];
        addFilter(filters, "AID", "EQUAL", params.accountId);
        addFilter(filters, "CO", "EQUAL", params.company);
        addFilter(filters, "DCT", "EQUAL", params.documentType);
        addFilter(filters, "FY", "EQUAL", params.fiscalYear);
        addFilter(filters, "PN", "EQUAL", params.period);
        if (params.dateFrom) addFilter(filters, "DGJ", "GREATER_EQUAL", params.dateFrom);
        if (params.dateTo) addFilter(filters, "DGJ", "LESS_EQUAL", params.dateTo);

        const response = await queryTable({
          tableName: "F0901",
          columns: ["AID", "DOC", "DCT", "KCO", "CO", "MCU", "LT", "FY", "PN", "AA", "DGJ", "EXR", "AN8", "JELN"],
          filters: filters.length > 0 ? filters : undefined,
          maxRows: params.maxRows,
        });

        const gridData = response.fs_DATABROWSE?.data?.gridData;
        if (!gridData) {
          return {
            content: [{ type: "text" as const, text: "No journal entries found matching the criteria." }],
          };
        }

        const output = {
          tableName: "F0901",
          description: "GL Journal Entry Detail",
          totalRecords: gridData.summary.records,
          returnedRecords: gridData.rowset.length,
          hasMore: gridData.summary.moreRecords,
          rows: gridData.rowset,
        };

        return {
          content: [{ type: "text" as const, text: truncate(JSON.stringify(output, null, 2), gridData.summary.records) }],
        };
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        return {
          isError: true,
          content: [{ type: "text" as const, text: `Error querying F0901: ${msg}` }],
        };
      }
    }
  );
}

// ──────────────────────────────────────────────────────────────
// Tool 4: A/P to G/L Integrity Check (programmatic R047001A)
// ──────────────────────────────────────────────────────────────

export function registerApGlIntegrityCheck(server: McpServer): void {
  server.registerTool(
    "jde_ap_gl_integrity_check",
    {
      title: "A/P to G/L Integrity Check",
      description: `Perform a programmatic A/P to G/L integrity check — the same logic as JDE report R047001A.

This tool:
1. Queries F0411 (A/P Ledger) and sums voucher amounts grouped by GL offset code (GLPT)
2. Queries F0902 (Account Balances) for the corresponding GL accounts
3. Compares the AP subledger totals against the GL balance for each offset account
4. Returns matches (balanced) and discrepancies (out of balance)

This is the KEY tool for R047001A analysis. Use it to verify whether the PDF report's
findings are consistent with the current state of the JDE data.

Args:
  - company: Company code (CO). Required.
  - fiscalYear: Fiscal year (FY). Required.
  - periodFrom: Starting period (PN). Required.
  - periodTo: Ending period (PN). Required.
  - glOffset: Specific GLPT to check. Omit for ALL offset accounts.

Returns:
  { summary, matches, discrepancies } where each entry shows the GLPT code,
  AP subtotal, GL balance, and the difference.`,
      inputSchema: JdeApGlIntegrityCheckSchema,
      annotations: {
        readOnlyHint: true,
        destructiveHint: false,
        idempotentHint: true,
        openWorldHint: true,
      },
    },
    async (params: JdeApGlIntegrityCheckInput) => {
      try {
        // ── Step 1: Query F0411 for AP voucher amounts ──
        const apFilters: FilterDef[] = [];
        addFilter(apFilters, "CO", "EQUAL", params.company);
        addFilter(apFilters, "FY", "EQUAL", params.fiscalYear);
        addFilter(apFilters, "PN", "GREATER_EQUAL", params.periodFrom);
        addFilter(apFilters, "PN", "LESS_EQUAL", params.periodTo);
        if (params.glOffset) addFilter(apFilters, "GLPT", "EQUAL", params.glOffset);

        const apResponse = await queryTable({
          tableName: "F0411",
          columns: ["GLPT", "AG", "AAP", "CO", "FY", "PN", "AID"],
          filters: apFilters,
          maxRows: 500,
        });

        const apRows = apResponse.fs_DATABROWSE?.data?.gridData?.rowset ?? [];

        // Group AP amounts by GLPT
        const apByGlpt: Record<string, { gross: number; open: number; count: number; aids: Set<string> }> = {};
        for (const row of apRows) {
          const glpt = String(row["F0411_GLPT"] ?? row["GLPT"] ?? "UNKNOWN");
          const gross = Number(row["F0411_AG"] ?? row["AG"] ?? 0);
          const open = Number(row["F0411_AAP"] ?? row["AAP"] ?? 0);
          const aid = String(row["F0411_AID"] ?? row["AID"] ?? "");

          if (!apByGlpt[glpt]) {
            apByGlpt[glpt] = { gross: 0, open: 0, count: 0, aids: new Set() };
          }
          apByGlpt[glpt].gross += gross;
          apByGlpt[glpt].open += open;
          apByGlpt[glpt].count += 1;
          if (aid) apByGlpt[glpt].aids.add(aid);
        }

        // ── Step 2: Query F0902 for GL balances ──
        const glFilters: FilterDef[] = [];
        addFilter(glFilters, "CO", "EQUAL", params.company);
        addFilter(glFilters, "LT", "EQUAL", "AA");
        addFilter(glFilters, "FY", "EQUAL", params.fiscalYear);

        const glResponse = await queryTable({
          tableName: "F0902",
          columns: ["AID", "CO", "OBJ", "SUB", "FY",
                     "AN01", "AN02", "AN03", "AN04", "AN05", "AN06",
                     "AN07", "AN08", "AN09", "AN10", "AN11", "AN12",
                     "AN13", "AN14", "BORG"],
          filters: glFilters,
          maxRows: 500,
        });

        const glRows = glResponse.fs_DATABROWSE?.data?.gridData?.rowset ?? [];

        // Build a map of AID → sum of period amounts for requested range
        const glByAid: Record<string, number> = {};
        for (const row of glRows) {
          const aid = String(row["F0902_AID"] ?? row["AID"] ?? "");
          let periodSum = 0;
          for (let p = params.periodFrom; p <= params.periodTo; p++) {
            const key = `AN${String(p).padStart(2, "0")}`;
            const prefixedKey = `F0902_${key}`;
            periodSum += Number(row[prefixedKey] ?? row[key] ?? 0);
          }
          glByAid[aid] = (glByAid[aid] ?? 0) + periodSum;
        }

        // ── Step 3: Compare AP vs GL ──
        const matches: Array<Record<string, unknown>> = [];
        const discrepancies: Array<Record<string, unknown>> = [];

        for (const [glpt, ap] of Object.entries(apByGlpt)) {
          // Find GL balance for any AID referenced by this GLPT group
          let glTotal = 0;
          const matchedAids: string[] = [];
          for (const aid of ap.aids) {
            if (glByAid[aid] !== undefined) {
              glTotal += glByAid[aid];
              matchedAids.push(aid);
            }
          }

          const difference = Math.round((ap.gross - glTotal) * 100) / 100;
          const entry = {
            glOffset: glpt,
            apGrossTotal: Math.round(ap.gross * 100) / 100,
            apOpenTotal: Math.round(ap.open * 100) / 100,
            glPeriodTotal: Math.round(glTotal * 100) / 100,
            difference,
            voucherCount: ap.count,
            matchedAccountIds: matchedAids,
          };

          if (Math.abs(difference) < 0.01) {
            matches.push(entry);
          } else {
            discrepancies.push(entry);
          }
        }

        const output = {
          reportName: "A/P to G/L Integrity Check (R047001A)",
          company: params.company,
          fiscalYear: params.fiscalYear,
          periodRange: `${params.periodFrom}-${params.periodTo}`,
          glOffsetFilter: params.glOffset ?? "ALL",
          apVouchersAnalyzed: apRows.length,
          glAccountsAnalyzed: glRows.length,
          summary: {
            totalOffsetGroups: Object.keys(apByGlpt).length,
            balanced: matches.length,
            discrepancies: discrepancies.length,
          },
          discrepancies,
          matches,
        };

        return {
          content: [{ type: "text" as const, text: truncate(JSON.stringify(output, null, 2)) }],
        };
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        return {
          isError: true,
          content: [{
            type: "text" as const,
            text: `Error running integrity check: ${msg}. Verify company '${params.company}' and fiscal year ${params.fiscalYear} exist in F0411.`,
          }],
        };
      }
    }
  );
}

File: jde-mcp-python/tools/integrity.py
"""Layer 2 — AP/GL Integrity tools (programmatic R047001A)."""
import json
import logging
from typing import Optional
from mcp.server.fastmcp import FastMCP

from config import CHARACTER_LIMIT
from services.ais_client import query_table, add_filter

logger = logging.getLogger(__name__)


def _truncate(text: str, records: int | None = None) -> str:
    if len(text) <= CHARACTER_LIMIT:
        return text
    suffix = f"\n\n... [TRUNCATED — {records if records is not None else 'many'} total records. Narrow your filters.]"
    return text[:CHARACTER_LIMIT] + suffix


def register(mcp: FastMCP) -> None:

    @mcp.tool(
        name="jde_ap_voucher_query",
        description=(
            "Query the JDE Accounts Payable Ledger (F0411) for voucher pay items. "
            "Essential for R047001A integrity analysis. The GLPT (GL Offset) field is the primary grouping key. "
            "Key columns: DOC, DCT, KCO, SFX, CO, AN8, MCU, AG (gross), AAP (open), GLPT, AID, FY, PN, DGJ, PST, VINV. "
            "Args: company, supplier_number, document_type, pay_status (A/P/V), gl_offset, "
            "date_from (YYYY-MM-DD), date_to, fiscal_year, period, max_rows (default 100)."
        ),
        annotations={"readOnlyHint": True, "destructiveHint": False, "idempotentHint": True, "openWorldHint": True},
    )
    async def jde_ap_voucher_query(
        company: Optional[str] = None,
        supplier_number: Optional[str] = None,
        document_type: Optional[str] = None,
        pay_status: Optional[str] = None,
        gl_offset: Optional[str] = None,
        date_from: Optional[str] = None,
        date_to: Optional[str] = None,
        fiscal_year: Optional[str] = None,
        period: Optional[str] = None,
        max_rows: int = 100,
    ) -> str:
        try:
            filters: list[dict] = []
            add_filter(filters, "CO", "EQUAL", company)
            add_filter(filters, "AN8", "EQUAL", supplier_number)
            add_filter(filters, "DCT", "EQUAL", document_type)
            add_filter(filters, "PST", "EQUAL", pay_status)
            add_filter(filters, "GLPT", "EQUAL", gl_offset)
            add_filter(filters, "FY", "EQUAL", fiscal_year)
            add_filter(filters, "PN", "EQUAL", period)
            if date_from:
                add_filter(filters, "DGJ", "GREATER_EQUAL", date_from)
            if date_to:
                add_filter(filters, "DGJ", "LESS_EQUAL", date_to)

            resp = await query_table(
                table_name="F0411",
                columns=["DOC", "DCT", "KCO", "SFX", "CO", "AN8", "MCU", "AG", "AAP", "PAAP", "GLPT", "AID", "FY", "PN", "DGJ", "PST", "VINV"],
                filters=filters or None,
                max_rows=max_rows,
            )
            grid = (resp.get("fs_DATABROWSE") or {}).get("data", {}).get("gridData")
            if not grid:
                return "No vouchers found matching the criteria."

            output = {
                "tableName": "F0411",
                "description": "A/P Voucher Pay Items",
                "totalRecords": grid.get("summary", {}).get("records"),
                "returnedRecords": len(grid.get("rowset", [])),
                "hasMore": grid.get("summary", {}).get("moreRecords"),
                "rows": grid.get("rowset", []),
            }
            return _truncate(json.dumps(output, indent=2), grid.get("summary", {}).get("records"))
        except Exception as exc:
            raise RuntimeError(f"Error querying F0411: {exc}") from exc

    @mcp.tool(
        name="jde_gl_balance_query",
        description=(
            "Query the JDE Account Balances table (F0902) for period-level GL balances. "
            "Returns balances by period for each account/ledger type/fiscal year combination. "
            "Key columns: AID, CO, MCU, OBJ, SUB, LT, FY, CTRY, AN01-AN14 (period amounts), BORG. "
            "Args: company, business_unit (MCU), object_account (OBJ) e.g. '1110', subsidiary, "
            "ledger_type (default 'AA'), fiscal_year, max_rows (default 100)."
        ),
        annotations={"readOnlyHint": True, "destructiveHint": False, "idempotentHint": True, "openWorldHint": True},
    )
    async def jde_gl_balance_query(
        company: Optional[str] = None,
        business_unit: Optional[str] = None,
        object_account: Optional[str] = None,
        subsidiary: Optional[str] = None,
        ledger_type: Optional[str] = None,
        fiscal_year: Optional[str] = None,
        max_rows: int = 100,
    ) -> str:
        try:
            filters: list[dict] = []
            add_filter(filters, "CO", "EQUAL", company)
            add_filter(filters, "MCU", "EQUAL", business_unit)
            add_filter(filters, "OBJ", "EQUAL", object_account)
            add_filter(filters, "SUB", "EQUAL", subsidiary)
            add_filter(filters, "LT", "EQUAL", ledger_type)
            add_filter(filters, "FY", "EQUAL", fiscal_year)

            period_cols = [f"AN{p:02d}" for p in range(1, 15)]
            resp = await query_table(
                table_name="F0902",
                columns=["AID", "CO", "MCU", "OBJ", "SUB", "LT", "FY", "CTRY"] + period_cols + ["BORG"],
                filters=filters or None,
                max_rows=max_rows,
            )
            grid = (resp.get("fs_DATABROWSE") or {}).get("data", {}).get("gridData")
            if not grid:
                return "No account balances found matching the criteria."

            output = {
                "tableName": "F0902",
                "description": "Account Balances by Period",
                "totalRecords": grid.get("summary", {}).get("records"),
                "returnedRecords": len(grid.get("rowset", [])),
                "hasMore": grid.get("summary", {}).get("moreRecords"),
                "rows": grid.get("rowset", []),
            }
            return _truncate(json.dumps(output, indent=2), grid.get("summary", {}).get("records"))
        except Exception as exc:
            raise RuntimeError(f"Error querying F0902: {exc}") from exc

    @mcp.tool(
        name="jde_gl_detail_query",
        description=(
            "Query the JDE Account Ledger (F0901) for individual GL journal entries. "
            "Use for drill-down when an integrity discrepancy is found. "
            "Key columns: AID, DOC, DCT, KCO, CO, MCU, LT, FY, PN, AA (amount), DGJ, EXR, AN8, JELN. "
            "Args: account_id (AID), company, document_type, fiscal_year, period, "
            "date_from (YYYY-MM-DD), date_to, max_rows (default 100)."
        ),
        annotations={"readOnlyHint": True, "destructiveHint": False, "idempotentHint": True, "openWorldHint": True},
    )
    async def jde_gl_detail_query(
        account_id: Optional[str] = None,
        company: Optional[str] = None,
        document_type: Optional[str] = None,
        fiscal_year: Optional[str] = None,
        period: Optional[str] = None,
        date_from: Optional[str] = None,
        date_to: Optional[str] = None,
        max_rows: int = 100,
    ) -> str:
        try:
            filters: list[dict] = []
            add_filter(filters, "AID", "EQUAL", account_id)
            add_filter(filters, "CO", "EQUAL", company)
            add_filter(filters, "DCT", "EQUAL", document_type)
            add_filter(filters, "FY", "EQUAL", fiscal_year)
            add_filter(filters, "PN", "EQUAL", period)
            if date_from:
                add_filter(filters, "DGJ", "GREATER_EQUAL", date_from)
            if date_to:
                add_filter(filters, "DGJ", "LESS_EQUAL", date_to)

            resp = await query_table(
                table_name="F0901",
                columns=["AID", "DOC", "DCT", "KCO", "CO", "MCU", "LT", "FY", "PN", "AA", "DGJ", "EXR", "AN8", "JELN"],
                filters=filters or None,
                max_rows=max_rows,
            )
            grid = (resp.get("fs_DATABROWSE") or {}).get("data", {}).get("gridData")
            if not grid:
                return "No journal entries found matching the criteria."

            output = {
                "tableName": "F0901",
                "description": "GL Journal Entry Detail",
                "totalRecords": grid.get("summary", {}).get("records"),
                "returnedRecords": len(grid.get("rowset", [])),
                "hasMore": grid.get("summary", {}).get("moreRecords"),
                "rows": grid.get("rowset", []),
            }
            return _truncate(json.dumps(output, indent=2), grid.get("summary", {}).get("records"))
        except Exception as exc:
            raise RuntimeError(f"Error querying F0901: {exc}") from exc

    @mcp.tool(
        name="jde_ap_gl_integrity_check",
        description=(
            "Perform a programmatic A/P to G/L integrity check — same logic as JDE report R047001A. "
            "1. Queries F0411 and sums voucher amounts grouped by GL offset code (GLPT). "
            "2. Queries F0902 for corresponding GL accounts. "
            "3. Compares AP subledger totals against GL balance — identifies matches and discrepancies. "
            "Args: company (required), fiscal_year (required), period_from (required), period_to (required), "
            "gl_offset (optional, omit for ALL offset accounts). "
            "Returns: { summary, matches, discrepancies } each with glOffset, AP total, GL balance, and difference."
        ),
        annotations={"readOnlyHint": True, "destructiveHint": False, "idempotentHint": True, "openWorldHint": True},
    )
    async def jde_ap_gl_integrity_check(
        company: str,
        fiscal_year: str,
        period_from: int,
        period_to: int,
        gl_offset: Optional[str] = None,
    ) -> str:
        try:
            # Step 1: query F0411
            ap_filters: list[dict] = []
            add_filter(ap_filters, "CO", "EQUAL", company)
            add_filter(ap_filters, "FY", "EQUAL", fiscal_year)
            add_filter(ap_filters, "PN", "GREATER_EQUAL", str(period_from))
            add_filter(ap_filters, "PN", "LESS_EQUAL", str(period_to))
            if gl_offset:
                add_filter(ap_filters, "GLPT", "EQUAL", gl_offset)

            ap_resp = await query_table(
                table_name="F0411",
                columns=["GLPT", "AG", "AAP", "CO", "FY", "PN", "AID"],
                filters=ap_filters,
                max_rows=500,
            )
            ap_rows = (ap_resp.get("fs_DATABROWSE") or {}).get("data", {}).get("gridData", {}).get("rowset", [])

            # Group AP by GLPT
            ap_by_glpt: dict[str, dict] = {}
            for row in ap_rows:
                glpt = str(row.get("F0411_GLPT") or row.get("GLPT") or "UNKNOWN")
                gross = float(row.get("F0411_AG") or row.get("AG") or 0)
                open_amt = float(row.get("F0411_AAP") or row.get("AAP") or 0)
                aid = str(row.get("F0411_AID") or row.get("AID") or "")
                if glpt not in ap_by_glpt:
                    ap_by_glpt[glpt] = {"gross": 0.0, "open": 0.0, "count": 0, "aids": set()}
                ap_by_glpt[glpt]["gross"] += gross
                ap_by_glpt[glpt]["open"] += open_amt
                ap_by_glpt[glpt]["count"] += 1
                if aid:
                    ap_by_glpt[glpt]["aids"].add(aid)

            # Step 2: query F0902
            gl_filters: list[dict] = []
            add_filter(gl_filters, "CO", "EQUAL", company)
            add_filter(gl_filters, "LT", "EQUAL", "AA")
            add_filter(gl_filters, "FY", "EQUAL", fiscal_year)

            period_cols = [f"AN{p:02d}" for p in range(1, 15)]
            gl_resp = await query_table(
                table_name="F0902",
                columns=["AID", "CO", "OBJ", "SUB", "FY"] + period_cols + ["BORG"],
                filters=gl_filters,
                max_rows=500,
            )
            gl_rows = (gl_resp.get("fs_DATABROWSE") or {}).get("data", {}).get("gridData", {}).get("rowset", [])

            # AID → sum of requested period amounts
            gl_by_aid: dict[str, float] = {}
            for row in gl_rows:
                aid = str(row.get("F0902_AID") or row.get("AID") or "")
                period_sum = 0.0
                for p in range(period_from, period_to + 1):
                    key = f"AN{p:02d}"
                    period_sum += float(row.get(f"F0902_{key}") or row.get(key) or 0)
                gl_by_aid[aid] = gl_by_aid.get(aid, 0.0) + period_sum

            # Step 3: compare
            matches: list[dict] = []
            discrepancies: list[dict] = []

            for glpt, ap in ap_by_glpt.items():
                gl_total = 0.0
                matched_aids: list[str] = []
                for aid in ap["aids"]:
                    if aid in gl_by_aid:
                        gl_total += gl_by_aid[aid]
                        matched_aids.append(aid)

                difference = round(ap["gross"] - gl_total, 2)
                entry = {
                    "glOffset": glpt,
                    "apGrossTotal": round(ap["gross"], 2),
                    "apOpenTotal": round(ap["open"], 2),
                    "glPeriodTotal": round(gl_total, 2),
                    "difference": difference,
                    "voucherCount": ap["count"],
                    "matchedAccountIds": matched_aids,
                }
                (matches if abs(difference) < 0.01 else discrepancies).append(entry)

            output = {
                "reportName": "A/P to G/L Integrity Check (R047001A)",
                "company": company,
                "fiscalYear": fiscal_year,
                "periodRange": f"{period_from}-{period_to}",
                "glOffsetFilter": gl_offset or "ALL",
                "apVouchersAnalyzed": len(ap_rows),
                "glAccountsAnalyzed": len(gl_rows),
                "summary": {
                    "totalOffsetGroups": len(ap_by_glpt),
                    "balanced": len(matches),
                    "discrepancies": len(discrepancies),
                },
                "discrepancies": discrepancies,
                "matches": matches,
            }
            return _truncate(json.dumps(output, indent=2))
        except Exception as exc:
            raise RuntimeError(
                f"Error running integrity check: {exc}. Verify company '{company}' and fiscal year {fiscal_year} exist in F0411."
            ) from exc
